content/docs/user-guide/assets/pipeline/asset-builders.md

```diff
@@ -13,28 +13,147 @@ The Open Asset Import Library supports [many scene formats](https://github.com/a
+
+* Source assets
+   * Source assets are files that exist in scan directories for an O3DE project, that match an Asset Builder's registered pattern, and result in a job being created when create jobs is called with the source asset for the Asset Builder.
+   * Read more about [source assets here.](/docs/user-guide/assets/pipeline/source-assets/)
+* Non-asset source files
+   * Non-asset source files are files that may or may not be in a scan directory, that are loaded and referenced when processing a source asset.
+   * The difference between a source asset and a non-asset source file is either:
+      * The file is not in a scan directory,
+      * The file does not match a file pattern for any asset builder descriptor, or
+      * The file does not generate a job when create jobs is called for the file for matching asset builders
+* Product assets
+   * Product assets are the run-time ready output of an asset processing job.
+* Intermediate assets
+   * Intermediate assets are source assets that are generated as a product of an asset processing job.
+   * Read more about [intermediate assets here.](/docs/user-guide/assets/pipeline/intermediate-assets/)
+
+#### References from source assets and non-asset source files to other files
+
+![Source asset references](/images/user-guide/assets/pipeline/asset_builders/source_asset_references.png)
+
+Source assets and non-asset source files can reference each other in many different ways, and each case may require unique handling at the time of authoring an asset builder.
+
+Source assets and non-asset source files may have these references:
+* source assets by UUID
+* source assets by a path
+* non-asset source files in a scan directory by a path
+* non-asset source files outside a scan directory by a path
+* product assets by asset ID
+* product assets by a path
+
+Path references from source asset and non-asset source files may be defined in these ways:
+* Relative to a scan directory
+* Relative to the asset with the reference
+* Absolute
+* Relative to a directory that may be difficult to infer from the contents or location of the source asset
+
+Source assets come in two forms, relative to how these references work:
+* A format that the builder author can modify and control.
+  * Example: A custom format defined for O3DE, like prefab files.
+* A format that the builder author has no control over.
+  * Example: A format defined outside of O3DE, like FBX files.
+
+When authoring a builder, if the format of the source asset is something the author has control over, then it's useful to manage how the source asset references other files.
+* When referencing a source asset or product asset, it's preferred to reference by UUID or Asset ID, because these will be more stable than a path. Read more about [asset identifiers here.](/docs/user-guide/assets/pipeline/asset-dependencies-and-identifiers/#asset-identifiers)
```

Comment:
Nit: Maybe we should force people to use the UUID/Asset ID to reference source/product asset. The APIs for resolving file path to asset ID are for the compatibility with the legacy cry system and should be avoided in any new O3DE systems.

Reply:
We can't force it because many source assets are going to be in formats we don't control. For example, an FBX file can have paths to textures baked in.